File: fiontb/spatial/_test/test_aabb.py
import unittest

import torch
import torch.testing as torcht

from fiontb.sparse.aabb import AABB, subdivide_octo


class TestAABB(unittest.TestCase):
    def test_subdivide_octo(self):
        box = AABB((-10, -10, -10), (10, 10, 10))

        boxes = subdivide_octo(box)

        self.assertEqual(8, len(boxes))
        # Left-Bottom-Near
        torcht.assert_allclose(boxes[0].min, torch.Tensor([-10., -10., -10.]))
        torcht.assert_allclose(boxes[0].max, torch.Tensor([0., 0., 0.]))

        # Left-Bottom-Far
        torcht.assert_allclose(boxes[1].min, torch.Tensor([-10., -10., 0.]))
        torcht.assert_allclose(boxes[1].max, torch.Tensor([0., 0., 10.]))

        # Left-Top-Far
        torcht.assert_allclose(boxes[2].min, torch.Tensor([-10., 0., 0.]))
        torcht.assert_allclose(boxes[2].max, torch.Tensor([0., 10., 10.]))

        # Left-Top-Near
        torcht.assert_allclose(boxes[3].min, torch.Tensor([-10., 0., -10.]))
        torcht.assert_allclose(boxes[3].max, torch.Tensor([0., 10., 0.]))

        # Right-Bottom-Near
        torcht.assert_allclose(boxes[4].min, torch.Tensor([0., -10., -10.]))
        torcht.assert_allclose(boxes[4].max, torch.Tensor([10., 0., 0.]))

        # Right-Bottom-Far
        torcht.assert_allclose(boxes[5].min, torch.Tensor([0., -10., 0.]))
        torcht.assert_allclose(boxes[5].max, torch.Tensor([10., 0., 10.]))

        # Right-Top-Far
        torcht.assert_allclose(boxes[6].min, torch.Tensor([0., 0., 0.]))
        torcht.assert_allclose(boxes[6].max, torch.Tensor([10., 10., 10.]))

        # Right-Top-Near
        torcht.assert_allclose(boxes[7].min, torch.Tensor([0., 0., -10.]))
        torcht.assert_allclose(boxes[7].max, torch.Tensor([10., 10., 0.]))

    def test_isinside(self):
        box = AABB((-10, -10, -10), (10, 10, 10))
        inside_mask = box.is_inside(
            torch.Tensor([[0, 0, 0],
                          [6, 7, 2],
                          [-10, -10, -10],
                          [-10, 0, 5],
                          [15, 14, 13],
                          [15, 8, 7],
                          [15, 8, 13],
                          [5, 12, 3],
                          [15, 10, 10]]))

        torcht.assert_allclose(torch.tensor([1, 1, 1, 1, 0, 0, 0, 0, 0], dtype=torch.uint8),
                               inside_mask)
        print(inside_mask)

File: src/test/test_aabb.cpp
#include "catch.hpp"

#include <fiontb/aabb.hpp>

TEST_CASE("AABB", "[aabb]") {
  fiontb::AABB box2(Eigen::Vector3f(2, 1, 3), Eigen::Vector3f(5, 4, 9));

  // Left half intersection
  CHECK(box2.Intersects(Eigen::Vector3f(1, 3, 6), Eigen::Vector3f(2.5, 1.5, 4),
                        Eigen::Vector3f(3.5, 3.5, 6)));

  // Left no intersection
  CHECK(!box2.Intersects(Eigen::Vector3f(1, 3, 6), Eigen::Vector3f(1.8, 1.5, 4),
                         Eigen::Vector3f(1.9, 3.5, 6)));

  // Right half intersection
  CHECK(box2.Intersects(Eigen::Vector3f(4, 3, 6), Eigen::Vector3f(5.5, 1.5, 4),
                        Eigen::Vector3f(7.5, 3.5, 6)));

  fiontb::AABB box3(Eigen::Vector3f(-0.0931465998, 0.107758701, 0.000578399748),
                    Eigen::Vector3f(-0.017493749, 0.181896999, 0.0578007996));
  CHECK(
      box3.IsInside(Eigen::Vector3f(-0.0235083941, 0.126511514, 0.0114857228)));

  Eigen::Vector3f p0(-0.00632139016, 0.0931651965, -0.0328935012);
  Eigen::Vector3f p1(-0.0144477999, 0.0879274011, -0.0380297005);
  Eigen::Vector3f p2(-0.0184118003, 0.0948058963, -0.0307433996);

  
}
